add network to Invoice

zoedberg opened the pull request:
As discussed on the last dev call this PR proposes to add an optional `chain` field to the invoice which purpose is to allow early detection of chain incompatibilities between sender and receiver.

zoedberg committed:
add network to Invoice

dr-orlovsky commented:
I am still not entirely convinced about this.

Chain is 32 bytes, which means + ~50 chars in Bech32 encoding and much more complex QR code fitting. At the same time it is impossible to do a payment on a wrong chain since the asset it is always chain-specific. I.e. the value proposition of this additional complexity is not clear for me.

zoedberg committed:
Update rust edition from 2018 to 2021

zoedberg commented:
I didn't notice that `Chain` was serialized with all its `ChainParams`. @dr-orlovsky you are right, the invoice would get too big in that case (e.g. from `i1q93kqcryunda8mxyd4cad0tzgaxfpc33e85e4fsej0370lpa7r8kmxupjt2y23sxvcrq3crqnqel2lpj3qzsqynq3xk` to `i1q93kqcryunda8mxyd4cad0tzgaxfpc33e85e4fsej0370lpa7r8kmxupjt2y2fsxvcrq3crqnqel2lpj3rzs69veclq2d988yc5v28ted274pqhna0506x58lkk79r94uj8klknw7wckxg3l8dp4z6n6f948rjd0l4dklsfe9sxynj29y44l84hyxypy6qtpqcjjdzy39sxz2n3pec7qcg2qv59yaxf9e8uej72steckyjt7w9rzds7r47kzp9vy4srf5zcqc20xc4`).

I changed this PR in order to add an `InvoiceChain` enum, much simpler than the `Chain` enum, that supports conversion from `Chain` to `InvoiceChain` for most used chains.

With this change, adding a chain to the invoice doesn't make it too big as before (e.g. from `i1q93kqcpuhrhu3e7yjsus7867jl83906e0djt8sw56vwdnm80mhhf7ufm6f60jz3rqvesxprsxpk85ap7r8zqyqqs9ltr3` to `i1q93kqcpuhrhu3e7yjsus7867jl83906e0djt8sw56vwdnm80mhhf7ufm6f60jzsnqvesxprsxpk85ap7r8zw9p52xqerqqgq62wj9w`).

This should cover most use cases. Since `chain` is an optional field and TLV is a very cool system, we can add this field without worrying too much, as users of this library will not see any size increase in their invoices when no chain is embedded. This should be a big usability improvement. As an example, users can create invoices that do not require a specific asset (e.g. for a token lottery) and scanning those invoices from wallets on different chains would be detected early, without the need for extra complexity elsewhere.

P.S. In order to implement `TryFrom` I've updated rust edition from 2018 to 2021

dr-orlovsky commented on rust/src/base.rs:
```diff
@@ -539,6 +552,64 @@ impl FromStr for ConsignmentEndpoint {
+
+    /// Default bitcoin regtest network
+    #[cfg_attr(feature = "serde", serde(rename = "regtest"))]
+    Regtest,
```
There is no such thing as regtest. Regtests are multiple, always private setup.

dr-orlovsky replied:
What I meant is that we shouldn't have this option here. You can't send a transaction to "a regtest" network without providing much more information (genesis, bitcoin core node address etc)

zoedberg replied:
Initially I was referring to bitcoin core default regtest, but there is no meaning in restricting invoices to just a regtest since that network will be used only for test purposes. I don't think there are use cases for regtest invoices outside test purposes so IMO a general `Regtest` enum is more than enough here, we don't need to differentiate different regtest networks in this context but it's still useful to have one

dr-orlovsky replied:
This doesn't make sense to me since the user might use different bitcoin core instance - with different regtest - than the invoice creator. In such cases it would be a situation which is really hard to debug. So I think we should have some regtest id here - maybe a u32-fingerprint of a specific regtest genesis or something

fedsten replied:
Invoices are meant to improve the UX of end users to detect mistakes before a state transition is attempted, they are not supposed to solve all the problems that developers testing in regtest may encounter. The worst case scenario here is that the invoice doesn't help the sender to understand he is on the wrong chain, and the send will fail, just like it happens right now with no invoices being used. The regtest id to me seems to add complexity to the invoice format with the scope of benefiting marginally some rare instances where developers are testing using multiple regtests.

dr-orlovsky commented on rust/src/base.rs:
```diff
@@ -539,6 +552,64 @@ impl FromStr for ConsignmentEndpoint {
+    serde(crate = "serde_crate")
+)]
+#[non_exhaustive]
+pub enum InvoiceChain {
```
I believe this should be not a `Chain` but rather a `Network`. The difference is the fact that you can have the same genesis (chain) but two different networks (Bitcoin, BCash etc). Also it is a network where the witness transaction is published, not a chain (to get into chain it has to be mined by miners).

And this brings the question why don't we use `bitcoin::Network` here.

zoedberg replied:
[`bitcoin::Network`](https://docs.rs/bitcoin/latest/bitcoin/network/constants/enum.Network.html) doesn't have the liquid network so I think we should keep our implementation, but I've renamed `InvoiceChain` to `Network` and the `chain` field to `network`

dr-orlovsky commented on rust/src/base.rs:
```diff
@@ -539,6 +552,64 @@ impl FromStr for ConsignmentEndpoint {
+#[non_exhaustive]
+pub enum InvoiceChain {
+    /// Bitcoin mainnet
+    #[cfg_attr(feature = "serde", serde(rename = "mainnet"))]
```
You do not need this if you add `#[serde(rename_all = "lowercase")]` at the structure level

zoedberg replied:
Thanks for the suggestion, I've updated the code

dr-orlovsky reviewed (changes requested):
The invoice can be an address-based. The address contains network flag. Also, xpub contains network flags - and they are used in descriptor-based invoices. Now we would have third place to put network information... I do not like this, we need a clear definition how the wallets should work with all these repeated (and potentially contradictory) information

zoedberg commented:
> The invoice can be an address-based. The address contains network flag. Also, xpub contains network flags - and they are used in descriptor-based invoices. Now we would have third place to put network information... I do not like this, we need a clear definition how the wallets should work with all these repeated (and potentially contradictory) information

I think this network flag doesn't require a clearer definition, mainly because handling of inconsistency is subjective (there is no right or wrong thing to do, it is a dev choice) and moreover the field is optional (it can be omitted if not needed, avoiding the extra complexity).

dr-orlovsky commented on rust/src/base.rs:
```diff
@@ -539,6 +552,53 @@ impl FromStr for ConsignmentEndpoint {
+    /// Bitcoin testnet version 3
+    Testnet3,
+
+    /// Bitcoin regtest network
```
```suggestion
    /// Regtest network used by an instance of Bitcoin Core which user wallet works with
```

dr-orlovsky commented:
@zoedberg I strictly against not having a standard way and leaving inconsistency in place in cases where we do _standard_ for monetary transaction invoices :D This is a clear concept NACK.

I understand that this inconsistency is already a part of the current invoice library, and is an issue unrelated to this PR. It has to be addressed separatedly.

However this PR adds to that inconsistency, spreading it from a case related just to descriptor/PBST-based payments to all forms of payments - with the aim of meeting some edge cases like "lottery".

My proposal on what to do with invoices, this problem etc: https://github.com/LNP-BP/LNPBPs/discussions/147

This actually addresses your concern and even "lottery" case much better. Thus, instead of introducing new fields with conflicting meaning I think we should move to a new invoicing format.

fedsten commented:
@dr-orlovsky your proposal looks interesting and for sure it is worth try to implement it in the future, but since invoicing is definitely lower priority compared to other RGB software, I think for now we can start using the current invoice format and evolve it later on. This PR, while not perfect, is still an improvement to the status quo and helps to improve the user experience for most cases, so I think it is worth merging it while a better invoicing format is being worked on and/or further improvements to the current invoices are being proposed.

dr-orlovsky commented:
This PR is not a "not perfect" but introduces more uncertainty about the way how the wallet should work - at a cost of marginal use case nobody ever used yet. So clear NACK from me,

If you'd like to add networking information to this invoice format, I recommend to work on aPR implementing this approach: https://github.com/orgs/LNP-BP/discussions/149

dr-orlovsky commented:
Superseded by new invoice format providing network information: https://github.com/RGB-WG/rgb-wallet/blob/master/src/invoice.rs#L80
